### PR TITLE
WebsocketManager backend

### PR DESCRIPTION
**Pull request by kafonek**
Adds a generic `WebsocketsManager` backend. Handles initial websocket creation and reconnecting. Offers hooks for authentication patterns and sending messages post-auth-reply. 

 - slight change to the abstract base so that subclasses can define `inbound_message_hook` and `outbound_message_hook` as class methods
 - new `-E websockets` and `-E all` install options

closes https://github.com/noteable-io/sending/pull/19

### CODE REVIEW

**Comment by kafonek** [comment]
Current dependencies on/for this PR:
* main
  * **PR #23** <a href="https://app.graphite.dev/github/pr/noteable-io/sending/23" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/noteable-io/sending/23?utm_source=stack-comment).

File: tests/test_asyncio.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import asyncio
```
Comment (tizz98) [inline tests/test_asyncio.py:1]: It seems strange to have tests testing `asyncio` here?

**Review by tizz98 — COMMENTED** [review]
looking good

File: tests/test_asyncio.py (hunk 1 above)
Comment (kafonek) [reply tests/test_asyncio.py:1]: Woops!! Yeah I still don't know why the one xfailed test isn't raising the `TimeoutError` like I believe it should. I was trying to reproduce that behavior in a minimal example, accidentally added and commited it.

**Review by tizz98 — DISMISSED** [review]

File: tests/test_websocket_backend.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    async def init_hook(mgr: WebsocketManager):
+        await mgr.send({"type": "unauthed_echo_request", "text": "Hello init_hook"})
+
+    json_manager.init_hook = init_hook
+    await json_manager.initialize()
+    reply = await run_until_message_type(json_manager, "unauthed_echo_reply")
+    assert reply == {"type": "unauthed_echo_reply", "text": "Hello init_hook"}
+
+
+@pytest.mark.xfail(reason="I don't know why this test doesn't work. Nick HALP")
```
Comment (nicholaswold) [inline tests/test_websocket_backend.py:99]: :joy:

File: sending/backends/websocket.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        # many reconnections.
+        self.max_reconnections = None
+
+        # Optional hooks that can be defined in a subclass or attached to an instance.
+        # First auth_hook then init_hook are called immediately after websocket connection
+        # They're separated into two hooks in case a subclass needs to authenticate, and
+        # also queue up initial messages that use .send() (which will wait until auth is accepted)
+        if not hasattr(self, "auth_hook"):
+            self.auth_hook: Optional[Callable] = None
+        self.init_hook: Optional[Callable] = None
```
Comment (nicholaswold) [inline sending/backends/websocket.py]: At some point I think we could graduate more hooks to the base manager. Could see some backends benefitting from the ability to define their own auth hook or init hook.

**Review by nicholaswold — APPROVED** [review]

**Pull request merged**